gateway/gateway_test.go

```diff
@@ -39,3 +42,198 @@ func TestValidate(t *testing.T) {
+	test.That(t, dev.AppKey, test.ShouldResemble, testAppKey)
+	test.That(t, dev.DevEui, test.ShouldResemble, testDevEUI)
+	test.That(t, dev.DecoderPath, test.ShouldResemble, "/newpath")
+
+	// Test remove device command
+	removeCmd := make(map[string]interface{})
+	removeCmd["remove_device"] = n.NodeName
+	doOverWire(s, removeCmd)
+	test.That(t, len(g.devices), test.ShouldEqual, 0)
+}
+
+func TestMergeNodes(t *testing.T) {
+	tests := []struct {
+		name      string
+		newNode   *node.Node
+		oldNode   *node.Node
+		want      *node.Node
+		expectErr error
+	}{
+		{
+			name: "OTAA join type - keep old Addr and AppSKey from join procedure",
+			newNode: &node.Node{
+				DecoderPath: "new/decoder/path",
+				NodeName:    "NewNode",
+				JoinType:    "OTAA",
+				AppKey:      testAppKey,
+				DevEui:      testDevEUI,
+				AppSKey: []byte{
+					0x11, 0x11, 0x40, 0x4C,
+					0x69, 0x6E, 0x6B, 0x4C,
+					0x6F, 0x52, 0x61, 0x32,
+					0x30, 0x31, 0x38, 0x23,
+				},
+			},
+			oldNode: &node.Node{
+				Addr:    testDeviceAddr,
+				AppSKey: testAppSKey,
+			},
+			want: &node.Node{
+				DecoderPath: "new/decoder/path",
+				NodeName:    "NewNode",
+				JoinType:    "OTAA",
+				Addr:        testDeviceAddr,
+				AppSKey:     testAppSKey,
+				AppKey:      testAppKey,
+				DevEui:      testDevEUI,
+			},
+			expectErr: nil,
+		},
+		{
+			name: "ABP join type - use new Addr and AppSKey",
+			newNode: &node.Node{
+				DecoderPath: "new/decoder/path",
+				NodeName:    "NewNode",
+				JoinType:    "ABP",
+				Addr:        testDeviceAddr,
+				AppSKey:     testAppSKey,
+			},
+			oldNode: &node.Node{
+				Addr: []byte{1, 5, 6, 7},
+				AppSKey: []byte{
+					0x11, 0x11, 0x40, 0x4C,
+					0x69, 0x6E, 0x6B, 0x4C,
+					0x6F, 0x52, 0x61, 0x32,
+					0x30, 0x31, 0x38, 0x23,
+				},
+			},
+			want: &node.Node{
+				DecoderPath: "new/decoder/path",
+				NodeName:    "NewNode",
+				JoinType:    "ABP",
+				Addr:        testDeviceAddr,
+				AppSKey:     testAppSKey,
+			},
+			expectErr: nil,
+		},
+		{
+			name: "Invalid join type",
+			newNode: &node.Node{
+				JoinType: "INVALID",
+			},
+			oldNode:   &node.Node{},
+			want:      nil,
+			expectErr: errUnexpectedJoinType,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got, err := mergeNodes(tt.newNode, tt.oldNode)
+			if tt.expectErr != nil {
+				test.That(t, err, test.ShouldBeError, tt.expectErr)
+			} else {
+				test.That(t, err, test.ShouldBeNil)
+				test.That(t, got, test.ShouldResemble, tt.want)
+			}
+		})
+	}
+}
+
+func TestUpdateReadings(t *testing.T) {
+	g := &Gateway{
+		lastReadings: make(map[string]interface{}),
+	}
+
+	// Test adding new readings
+	newReadings := map[string]interface{}{
+		"temp": 25.5,
+		"hum":  60,
+	}
+	g.updateReadings("device1", newReadings)
+	test.That(t, g.lastReadings["device1"], test.ShouldResemble, newReadings)
+
+	// Test updating existing readings
+	updatedReadings := map[string]interface{}{
+		"temp": 26.5,
+		"pres": 1013,
+	}
```

Comment:
This might benefit from a test case structure too.
```suggestion
	expected := map[string]interface{}{
		"temp": 26.5,
		"pres": 1013,
	}
```

Reply:
the expected is defined below this, maybe `newReading` if `updatedReadings` is confusing?

Reply:
How about a structure that looks like this so you're testing an API and the update readings fucntion, right not I guess it's okay since it's an input and output test but we can add more coverage within this test by perhaps doing:
```go
expected := some readings
g.updateReadings("device", expected)
readings, err := g.Readings(....)
test.That(t, err, test.ShouldBeNil)
test,That(t, readings, tets.ShouldResemble, expected)
```

Also are there any error cases we expect from update readings? We should test those too.

Reply:
The return of Readings() doesn't always equal the input to updateReadings, but I will combine the readings and updateReadings test. Theres not error cases.

Reply:
Actually I would prefer to keep the tests separate, I think its a lot cleaner to keep one test per function

Reply:
My  issue now is that there is no update readings -> last readings -> the actual API that is called: `Readings` test, so we would miss a case where something goes wrong where lastreadings is updated but that itsn't updated in the readings map. 

One test per function is cleaner, but make sure you're testing code paths that actually make sense to test, right now the Readings test does not tell me much about this driver's behaviour.

Reply:
Doesn't the `TestReadings ` below this test what you are referring to? The purpose of the Readings test below is to make sure that `g.lastReadings` and the readings map match.